slinky/runtime/depends_on.cc

```diff
@@ -416,4 +416,45 @@ var find_buffer_data_dependency(expr_ref e) {
   return result;
 }
 
+namespace {
+class side_effect_checker : public recursive_node_visitor {
+
+public:
+  bool has_side_effects = false;
+
+  void visit(const call* op) override {
+    switch (op->intrinsic) {
+    // Non-intrinsic could have a side-effect.
+    case intrinsic::none:
+    case intrinsic::semaphore_init:
+    case intrinsic::semaphore_signal:
+    case intrinsic::semaphore_wait:
+    case intrinsic::wait_for:
+    case intrinsic::trace_begin:
+    case intrinsic::trace_end:
+    case intrinsic::free: {
+      has_side_effects = true;
+      break;
```

Comment:
Just return early, no point in recursion here.

Also, no need for braces here.

Reply:
Done